src/Serilog.Sinks.EventGrid/Sinks/EventGrid/EventGridSink.cs

```diff
@@ -99,41 +99,67 @@ void GetEventInfoFromAttribute(CustomEventRequest customEvent)
-        var classAttribute = methods.FirstOrDefault(m => m.ReflectedType != null && m.ReflectedType.GetCustomAttribute<EventGridSinkAttribute>() != null)?.ReflectedType?.GetCustomAttribute<EventGridSinkAttribute>();
-        myCustomAttribute = classAttribute;
-      }
+      var subjectAttribute = GetSubjectFromAttribute(methods);
+      var typeAttribute = GetTypeFromAttribute(methods);
 
       // assign the event info, failing back to generic defaults
-      customEvent.Subject = customEvent.Subject ?? myCustomAttribute?.CustomEventSubject ?? GetSubject();
-      customEvent.EventType = customEvent.EventType ?? myCustomAttribute?.CustomEvenType ?? _customEventType ?? GetType();
+      customEvent.Subject = customEvent.Subject ?? subjectAttribute?.CustomEventSubject ?? GetSubject();
+      customEvent.EventType = customEvent.EventType ?? typeAttribute?.CustomEventType ?? _customEventType ?? GetEventType() ?? "AppDomain/Class";
 
       string GetSubject()
       {
-        // try to find the method where the log event initiated from
+        var methodName = GetMethodWithParams() ?? "Method/default";
+        return $"{methodName}";
+      }
+
+      string GetMethodWithParams()
+      {
+        var parameterNames = callingMethod != null && callingMethod.GetParameters().Any() 
+          ? callingMethod.GetParameters().Select(x => x.Name).Aggregate((combined, next) => combined += string.IsNullOrEmpty(combined) ? next : $"/{next}") 
+          : "default";
+        var methodWithParameters = $"{callingMethod.Name}/{parameterNames}";
+        return methodWithParameters;
+      }
+
+      string GetEventType()
+      {
         var assemblyName = callingMethod?.ReflectedType?.Assembly.GetName().Name ?? "General";
         var className = callingMethod?.ReflectedType?.Name ?? "Class";
-        var methodName = callingMethod.Name;
-        return $"{assemblyName}/{className}/{methodName}";
+        return $"{assemblyName}/{className}";
+      }
+    }
 
+    private EventGridTypeAttribute GetTypeFromAttribute(MethodBase[] methods)
+    {
+      EventGridTypeAttribute typeAttribute;
+      // look for the first method in the stack with the type attribute
+      var methodTypeAttribute = methods.FirstOrDefault(m => m.GetCustomAttribute<EventGridTypeAttribute>() != null)?.GetCustomAttribute<EventGridTypeAttribute>();
+      if (methodTypeAttribute != null)
+        typeAttribute = methodTypeAttribute;
+      else
+      {
+        // then look for the first class with the attribute, there can be only one
+        var classTypeAttribute = methods.FirstOrDefault(m => m.ReflectedType != null && m.ReflectedType.GetCustomAttribute<EventGridTypeAttribute>() != null)?.ReflectedType?.GetCustomAttribute<EventGridTypeAttribute>();
+        typeAttribute = classTypeAttribute;
       }
+      return typeAttribute;
+    }
 
-      string GetType()
+    private EventGridSubjectAttribute GetSubjectFromAttribute(MethodBase[] methods)
```

Comment:
looks like the subject function and type function are kind of same. can we merge this two into one function and pass T as generic type? something like   private T GetFromAttribute<T>(MethodBase[] methods)

Reply:
@jiaxiangl absolutely! good idea. I'll make the change